{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymongo\n",
    "import pprint\n",
    "import glob\n",
    "import pandas as pd\n",
    "\n",
    "db_connect = pymongo.MongoClient(\"localhost\", 27017)\n",
    "database_name = 'ye'\n",
    "db = db_connect[database_name]\n",
    "collection = db.list_collection_names(include_system_collections=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gr = db.genome_records"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genbank example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = gr.find_one({\"name\": \"NZ_CP010029.1\"})\n",
    "accession_id = g['refseq_accession_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['_id', 'name', 'species', 'description', 'sequence', 'assembly_name', 'biosample', 'bioproject', 'date', 'wgs_project', 'genome_coverage', 'organism', 'taxid', 'assembly_type', 'release_type', 'assembly_level', 'genome_representation', 'expected_final_version', 'excluded', 'genbank_accession_id', 'refseq_accession_id', 'r_g_identical', 'plasmid', 'present', 'hits', 'references', 'tags'])"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NZ_CP010029.1\n",
      "Yersinia entomophaga\n",
      "GCF_001656035.1\n"
     ]
    }
   ],
   "source": [
    "print (g['name'])\n",
    "print (g['species'])\n",
    "print (accession_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load in the feature table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>class</th>\n",
       "      <th>assembly</th>\n",
       "      <th>assembly_unit</th>\n",
       "      <th>seq_type</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>genomic_accession</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>product_accession</th>\n",
       "      <th>non-redundant_refseq</th>\n",
       "      <th>related_accession</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>feature_interval_length</th>\n",
       "      <th>product_length</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>GCF_001656035.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NZ_CP010029.1</td>\n",
       "      <td>24</td>\n",
       "      <td>1091</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PL78_RS20065</td>\n",
       "      <td>1068</td>\n",
       "      <td>NaN</td>\n",
       "      <td>partial;pseudo;old_locus_tag=PL78_00005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDS</td>\n",
       "      <td>without_protein</td>\n",
       "      <td>GCF_001656035.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NZ_CP010029.1</td>\n",
       "      <td>24</td>\n",
       "      <td>1091</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PL78_RS20065</td>\n",
       "      <td>1068</td>\n",
       "      <td>NaN</td>\n",
       "      <td>partial;pseudo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GCF_001656035.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NZ_CP010029.1</td>\n",
       "      <td>1352</td>\n",
       "      <td>1534</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PL78_RS00010</td>\n",
       "      <td>183</td>\n",
       "      <td>NaN</td>\n",
       "      <td>old_locus_tag=PL78_00010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDS</td>\n",
       "      <td>with_protein</td>\n",
       "      <td>GCF_001656035.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NZ_CP010029.1</td>\n",
       "      <td>1352</td>\n",
       "      <td>1534</td>\n",
       "      <td>-</td>\n",
       "      <td>WP_064512099.1</td>\n",
       "      <td>WP_064512099.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PL78_RS00010</td>\n",
       "      <td>183</td>\n",
       "      <td>60.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GCF_001656035.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NZ_CP010029.1</td>\n",
       "      <td>1700</td>\n",
       "      <td>1978</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PL78_RS00015</td>\n",
       "      <td>279</td>\n",
       "      <td>NaN</td>\n",
       "      <td>old_locus_tag=PL78_00015</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  feature            class         assembly     assembly_unit    seq_type  \\\n",
       "0    gene       pseudogene  GCF_001656035.1  Primary Assembly  chromosome   \n",
       "1     CDS  without_protein  GCF_001656035.1  Primary Assembly  chromosome   \n",
       "2    gene   protein_coding  GCF_001656035.1  Primary Assembly  chromosome   \n",
       "3     CDS     with_protein  GCF_001656035.1  Primary Assembly  chromosome   \n",
       "4    gene   protein_coding  GCF_001656035.1  Primary Assembly  chromosome   \n",
       "\n",
       "   chromosome genomic_accession  start   end strand product_accession  \\\n",
       "0         NaN     NZ_CP010029.1     24  1091      -               NaN   \n",
       "1         NaN     NZ_CP010029.1     24  1091      -               NaN   \n",
       "2         NaN     NZ_CP010029.1   1352  1534      -               NaN   \n",
       "3         NaN     NZ_CP010029.1   1352  1534      -    WP_064512099.1   \n",
       "4         NaN     NZ_CP010029.1   1700  1978      +               NaN   \n",
       "\n",
       "  non-redundant_refseq  related_accession                  name symbol  \\\n",
       "0                  NaN                NaN                   NaN    NaN   \n",
       "1                  NaN                NaN  hypothetical protein    NaN   \n",
       "2                  NaN                NaN                   NaN    NaN   \n",
       "3       WP_064512099.1                NaN  hypothetical protein    NaN   \n",
       "4                  NaN                NaN                   NaN    NaN   \n",
       "\n",
       "   GeneID     locus_tag  feature_interval_length  product_length  \\\n",
       "0     NaN  PL78_RS20065                     1068             NaN   \n",
       "1     NaN  PL78_RS20065                     1068             NaN   \n",
       "2     NaN  PL78_RS00010                      183             NaN   \n",
       "3     NaN  PL78_RS00010                      183            60.0   \n",
       "4     NaN  PL78_RS00015                      279             NaN   \n",
       "\n",
       "                                attributes  \n",
       "0  partial;pseudo;old_locus_tag=PL78_00005  \n",
       "1                           partial;pseudo  \n",
       "2                 old_locus_tag=PL78_00010  \n",
       "3                                      NaN  \n",
       "4                 old_locus_tag=PL78_00015  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filepath = \"/Users/gabefoley/Dropbox/PhD/Projects/Phylo_Island/2020/20201119_Testing_Feature_Tables/refseq/bacteria/\"\n",
    "feature_path = glob.glob(filepath + accession_id + \"/*_feature_table.txt.gz\")[0]\n",
    "df = pd.read_csv(feature_path, sep='\\t', compression='gzip')\n",
    "df.rename(columns = {'# feature':'feature'}, inplace = True)\n",
    "df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A1_expanded\n",
      "[798387, 801882]\n",
      "Found one\n",
      "toxin WP_064513225.1 798388 801882\n",
      "Found one\n",
      "hypothetical protein WP_064513227.1 801875 805969\n",
      "A2_expanded\n",
      "[801874, 805969]\n",
      "Found one\n",
      "toxin WP_064513225.1 798388 801882\n",
      "Found one\n",
      "hypothetical protein WP_064513227.1 801875 805969\n",
      "TcB_expanded\n",
      "[808162, 812626]\n",
      "Found one\n",
      "virulence protein WP_064513231.1 808163 812626\n",
      "TcC_expanded\n",
      "[815831, 818780]\n",
      "Found one\n",
      "hypothetical protein WP_084414279.1 815868 818780\n",
      "TcC_expanded\n",
      "[4177811, 4180709]\n",
      "Found one\n",
      "RHS repeat protein nan 4178646 4180709\n",
      "TcC_expanded\n",
      "[812721, 815646]\n",
      "Found one\n",
      "hypothetical protein WP_064513233.1 812722 815646\n",
      "TcdA1_expanded\n",
      "[801874, 805969]\n",
      "Found one\n",
      "toxin WP_064513225.1 798388 801882\n",
      "Found one\n",
      "hypothetical protein WP_064513227.1 801875 805969\n",
      "Chitinase_expanded\n",
      "[796661, 798302]\n",
      "Found one\n",
      "chitinase WP_064513223.1 796662 798302\n",
      "Chitinase_expanded\n",
      "[806066, 807968]\n",
      "Found one\n",
      "chitinase WP_064513229.1 806067 807968\n",
      "region1_expanded\n",
      "[801877, 802753]\n",
      "Found one\n",
      "toxin WP_064513225.1 798388 801882\n",
      "Found one\n",
      "hypothetical protein WP_064513227.1 801875 805969\n",
      "region1_expanded\n",
      "[803650, 805897]\n",
      "Found one\n",
      "hypothetical protein WP_064513227.1 801875 805969\n"
     ]
    }
   ],
   "source": [
    "for hit in g['hits']:\n",
    "    if 'expanded' in hit['region']:\n",
    "        print (hit['region'])\n",
    "        \n",
    "        genome_interval = pd.Interval(int(hit['start']), int(hit['end']), closed='both')\n",
    "        \n",
    "        print (genome_interval)\n",
    "                    \n",
    "        filtered = df.loc[df['assembly'] == g['refseq_accession_id']]\n",
    "        \n",
    "        filtered.head(5)\n",
    "\n",
    "        for gene in filtered.itertuples():\n",
    "            if gene.feature == 'CDS':\n",
    "                if not (gene.start > int(hit['end'])) and not (gene.end < int(hit['start'])):\n",
    "                    check_interval = pd.Interval(gene.start, gene.end, closed='both')\n",
    "\n",
    "                    if check_interval.overlaps(genome_interval):\n",
    "                        print ('Found one')\n",
    "                        print(gene.name, gene.product_accession, gene.start, gene.end)\n",
    "                    #\n",
    "                    # hit = models.Hits(object_id, record.name, new_reg, new_score, new_start, new_end,\n",
    "                    #                   expand,\n",
    "                    #                   strand, sequence)\n",
    "                    #\n",
    "                    # print('We are adding a refseq sequence in')\n",
    "                    #\n",
    "                    # print(hit.start)\n",
    "                    # print(hit.end)\n",
    "                    # curr.hits.append(hit)\n",
    "                    # curr.save()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>class</th>\n",
       "      <th>assembly</th>\n",
       "      <th>assembly_unit</th>\n",
       "      <th>seq_type</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>genomic_accession</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>product_accession</th>\n",
       "      <th>non-redundant_refseq</th>\n",
       "      <th>related_accession</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>feature_interval_length</th>\n",
       "      <th>product_length</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [feature, class, assembly, assembly_unit, seq_type, chromosome, genomic_accession, start, end, strand, product_accession, non-redundant_refseq, related_accession, name, symbol, GeneID, locus_tag, feature_interval_length, product_length, attributes]\n",
       "Index: []"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
