{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Variables to update can be found in checking_domains_setup.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from checking_domains_setup import *\n",
    "\n",
    "os.chdir(base_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Code to visualise phylogenetic tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ete3 import Tree, TreeStyle, TextFace, add_face_to_node, SeqMotifFace, NodeStyle, faces, ImgFace, CircleFace, AttrFace\n",
    "from IPython.display import display\n",
    "\n",
    "def get_domains(domains):\n",
    "    \n",
    "    pos_dict = {'RBD_A' : 0, 'RBD_C' : 1, 'RBD_B' : 2, 'NMD' : 3, 'RBD_D' : 4, 'TcB_BD_seed' : 5}\n",
    "\n",
    "    \n",
    "    domain_list = [\n",
    "        # seq.start, seq.end, shape, width, height, fgcolor, bgcolor\n",
    "        [10,  70, \"[]\", None, 20, \"black\", \"rgradient:lightgreen\", \"arial|3|black|RBD_A\"],\n",
    "        [80, 140, \"[]\", None, 20, \"black\", \"rgradient:blue\", \"arial|3|black|RBD_C\"],\n",
    "        [150, 210, \"[]\", None, 20, \"black\", \"rgradient:orange\", \"arial|3|black|RBD_B\"],\n",
    "        [220, 280, \"[]\", None, 20, \"black\", \"rgradient:purple\", \"arial|3|black|NMD\"],\n",
    "        [290, 350, \"[]\", None, 20, \"black\", \"rgradient:gray\", \"arial|3|black|RBD_D\"],\n",
    "        [360, 430, \"[]\", None, 20, \"black\", \"rgradient:darkgreen\", \"arial|3|black|TCB_BD\"]]\n",
    "    \n",
    "    \n",
    "    for k,v in pos_dict.items():\n",
    "        if k not in domains:\n",
    "            domain_list[v][6] = 'white'\n",
    "            domain_list[v][7] = \"arial|3|black|\"\n",
    "            \n",
    "    return domain_list\n",
    "\n",
    "\n",
    "def get_example_tree(tree, tag_dict, colour_dict, region_dict):\n",
    "            \n",
    "            colour = None\n",
    "\n",
    "            # Label all internal nodes\n",
    "            edge = 0\n",
    "            for node in tree.traverse():\n",
    "                if not node.is_leaf():\n",
    "                    node.name = \"N%d\" % edge\n",
    "                    edge += 1\n",
    "\n",
    "            # Get the colours for each extant genome\n",
    "            for node in tree.iter_descendants(\"postorder\"):\n",
    "                if node.is_leaf():\n",
    "                    long_name = node.name.split(\"_joined\")[0]\n",
    "                    short_name = node.name.split(\"_information\")[0]\n",
    "                    \n",
    "                    if long_name in tag_dict:\n",
    "                        colour = colour_dict[tag_dict[long_name][0]]\n",
    "                    elif short_name in tag_dict:\n",
    "                        colour = colour_dict[tag_dict[short_name][0]]\n",
    "                    else:\n",
    "                        colour = 'black'\n",
    "                    spaced_name = \" \".join(node.name.split(\"_\")[3:5])\n",
    "\n",
    "                    nameFace = TextFace(\"  \" + spaced_name, fsize=15, fgcolor='black')\n",
    "                    \n",
    "                    box_domains = get_domains([x for x in region_dict[node.name].keys()])\n",
    "                    \n",
    "                    seqFace = SeqMotifFace(seq=None, motifs=box_domains, gap_format=\"line\")\n",
    "                    node.add_face(seqFace, 0, \"aligned\")\n",
    "\n",
    "                    node.add_face(nameFace, column=0)\n",
    "\n",
    "                else:\n",
    "                    colour = 'black'\n",
    "                    \n",
    "                if colour == None:\n",
    "                    colour = 'black'\n",
    "\n",
    "                nstyle = NodeStyle()\n",
    "                nstyle[\"fgcolor\"] = colour\n",
    "                nstyle[\"size\"] = 20\n",
    "                node.set_style(nstyle)\n",
    "                \n",
    "            \n",
    "            \n",
    "                ts = TreeStyle()\n",
    "                ts.show_leaf_name = False\n",
    "                ts.branch_vertical_margin = 10\n",
    "\n",
    "                # if custom_layout:\n",
    "                #     ts.layout_fn = layout\n",
    "                #     ts.show_leaf_name = False\n",
    "\n",
    "            # ts.mode = \"c\"\n",
    "            ts.root_opening_factor = 1\n",
    "\n",
    "            return tree, ts\n",
    "\n",
    "def colour_tips(tree, tag_dict, colour_dict, region_dict, outpath=None, custom_layout=False):\n",
    "\n",
    "    tree, ts = get_example_tree(tree, tag_dict, colour_dict, region_dict)\n",
    "    if outpath:\n",
    "        tree.render(outpath, dpi=300, tree_style=ts)\n",
    "\n",
    "    display(tree.render('%%inline', dpi=300, tree_style=ts))\n",
    "    return tree, ts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Code to run the alignment, tree inference, domain searching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_profile_dict(base_dir, profile_name):\n",
    "    \n",
    "    print (f'Profile name is {profile_name}\\n')\n",
    "\n",
    "    profile_dir = f'{base_dir}/profiles/{profile_name}'\n",
    "    \n",
    "    print (f'Profile dir is {profile_dir}\\n')\n",
    "    profile_file = glob.glob(f'{profile_dir}/*.p')[0]\n",
    "\n",
    "    # Load the profile dict\n",
    "    with open(profile_file, 'rb') as file_path:\n",
    "        profile_dict = pickle.load(file_path)\n",
    "    \n",
    "    return profile_dict\n",
    "\n",
    "\n",
    "def setup_working_dir(base_dir, filename):\n",
    "    # Creating a working directory\n",
    "    if not os.path.isdir(f'{base_dir}/results/{filename}'):\n",
    "        print (\"Making directory\\n\")\n",
    "            \n",
    "        # Make the directory\n",
    "        os.mkdir(f'{base_dir}/results/{filename}')\n",
    "        \n",
    "        # Copy the FASTA file into the working directory\n",
    "        os.popen(f'cp {filename}.fasta ./results/{filename}/{filename}.fasta ')\n",
    "        \n",
    "        while not os.path.isfile(f'./results/{filename}/{filename}.fasta'):\n",
    "            os.wait()\n",
    "\n",
    "    file_dir = f'{base_dir}/results/{filename}/'\n",
    "\n",
    "    return file_dir\n",
    "\n",
    "def make_alignment(filename):\n",
    "    \n",
    "    # Make an alignment\n",
    "    if not os.path.isfile(filename + '.aln'):\n",
    "        print (\"Creating alignment\\n\")\n",
    "        while not os.path.isfile(f'{filename}.fasta'):\n",
    "            os.wait()\n",
    "        stdoutdata = subprocess.getoutput(f'mafft --genafpair --maxiterate 1000 --reorder {file_dir}{filename}.fasta > {filename}.aln')\n",
    "\n",
    "        print (\"Alignment created\\n\")\n",
    "        \n",
    "def make_tree(filename):\n",
    "    \n",
    "    # Make a tree\n",
    "    if not os.path.isfile(f'RAXML_bestTree.{filename}.nwk'):\n",
    "        print (\"Creating tree\\n\")\n",
    "\n",
    "        stdoutdata = subprocess.getoutput(f'raxml -m PROTGAMMAJTT -p 23456 -n {filename}.nwk -s {filename}.aln')\n",
    "\n",
    "        print (\"Tree created\\n\")\n",
    "\n",
    "def place_outgroup(filename, outgroup):\n",
    "    # Place the outgroup\n",
    "\n",
    "    if not os.path.isfile(f'{filename}.nwk'):\n",
    "        print (\"Placing outgroup\\n\")\n",
    "\n",
    "        tree = Tree(f'RAXML_bestTree.{filename}.nwk', format=1)\n",
    "        tree.set_outgroup(outgroup)\n",
    "\n",
    "        tree.write(outfile= f'{filename}.nwk', format=1)\n",
    "\n",
    "def search_for_domains(base_dir, file_dir, profile_name):\n",
    "    print (\"Searching for domains \\n\")\n",
    "    input_folder, output_folder, profile_folder, tree_folder = setup_domain_directories(base_dir, file_dir, profile_name)\n",
    "    region_dict = search_seqs_with_profiles(input_folder, output_folder, profile_folder, tree_folder)\n",
    "    region_dict, domain_dict = process_hmmer_results(region_dict, output_folder)\n",
    "    \n",
    "#     for path in [output_folder + '/*.output', input_folder + 'tmp/*' ]:\n",
    "#         remove_temp_files(path)\n",
    "    \n",
    "    return region_dict, domain_dict\n",
    "\n",
    "    \n",
    "def setup_domain_directories(base_dir, file_dir, profile_name):\n",
    "    for dir_name in ['domain_input', 'domain_output', 'tree_output']:\n",
    "        if not os.path.isdir(dir_name):\n",
    "            os.mkdir(dir_name)\n",
    "    \n",
    "    print ('Setup the domain directories \\n')\n",
    "    \n",
    "    input_folder = f'{file_dir}domain_input/'\n",
    "    output_folder = f'{file_dir}domain_output/'\n",
    "    profile_folder = f'{file_dir}{profile_name}/'\n",
    "    tree_folder = f'{file_dir}tree_output/'\n",
    "    \n",
    "    \n",
    "\n",
    "    shutil.copy2(f'{file_dir}{filename}.fasta', f'{input_folder}{filename}.fasta')\n",
    "    if not os.path.isdir(f'{profile_folder}'):\n",
    "        shutil.copytree(f'{base_dir}/profiles/{profile_name}', f'{profile_folder}')\n",
    "    \n",
    "\n",
    "    while not os.path.isdir(f'{file_dir}{profile_name}'):\n",
    "        os.wait()\n",
    "    \n",
    "    return input_folder, output_folder, profile_folder, tree_folder\n",
    "\n",
    "def search_seqs_with_profiles(input_folder, output_folder, profile_folder, tree_folder):\n",
    "    \n",
    "    print ('Searching the sequences with the profiles\\n')\n",
    "    domScore = 1\n",
    "\n",
    "    region_dict = {}\n",
    "\n",
    "    files = [x for x in os.listdir(input_folder) if x != \".DS_Store\" and x != 'tmp']\n",
    "    profiles = glob.glob(profile_folder + \"/*.hmm\")\n",
    "    \n",
    "    for file in files:\n",
    "\n",
    "        seqs = sequence.readFastaFile(input_folder + file)\n",
    "\n",
    "        # Write out each region in the file to a separate file\n",
    "\n",
    "        if not os.path.isdir(input_folder + \"tmp\"):\n",
    "            os.mkdir(input_folder + \"tmp\")\n",
    "\n",
    "        for seq in seqs:\n",
    "            sequence.writeFastaFile(input_folder + \"tmp/\" + seq.name, [seq])\n",
    "\n",
    "            if seq.name not in region_dict:\n",
    "                region_dict[seq.name] = {}\n",
    "\n",
    "        # Check each region with each profile\n",
    "            for profile in profiles:\n",
    "                profile_split =  profile.split(\"/\")[-1]\n",
    "                stdoutdata = subprocess.getoutput(f'hmmsearch -o {output_folder}{seq.name}_profile={profile_split}.output --domT {domScore} {profile_name}/{profile_split} {input_folder}tmp/{seq.name}')\n",
    "        \n",
    "    return region_dict\n",
    "\n",
    "def process_hmmer_results(region_dict, output_folder):\n",
    "    domain_dict = {}\n",
    "    \n",
    "    for infile in glob.glob(output_folder + '/*.output'):\n",
    "\n",
    "        seq_name = infile.split(output_folder)[1].split(\"_profile=\")[0]\n",
    "\n",
    "        qresult = SearchIO.read(infile, 'hmmer3-text')\n",
    "        if len(qresult.hits) > 0:\n",
    "            hsp = qresult[0][0]\n",
    "            domain = infile.split(\"profile=\")[1].split(\".\")[0]\n",
    "            start = hsp.hit_start\n",
    "            end = hsp.hit_end\n",
    "            region_dict[seq_name][domain] = (start, end)\n",
    "\n",
    "            if domain not in domain_dict:\n",
    "                domain_dict[domain] = []\n",
    "            domain_dict[domain].append(seq_name)\n",
    "    return region_dict, domain_dict\n",
    "\n",
    "\n",
    "\n",
    "def check_for_domains_to_add(profile_dict, domain_dict):\n",
    "\n",
    "    domains_to_add = defaultdict(list)\n",
    "    for domain, seq_list in profile_dict.items():\n",
    "        if not seq_list[0] == 'Pfam_seed':\n",
    "            if domain in domain_dict.keys():\n",
    "                for found_domain in domain_dict[domain]:\n",
    "                    if found_domain not in profile_dict[domain]:\n",
    "                        domains_to_add[domain].append(found_domain)\n",
    "    \n",
    "    return domains_to_add\n",
    "\n",
    "def add_domains(domains_to_add, region_dict, profile_dict, profile_name):\n",
    "        \n",
    "    for domain_name, to_add in domains_to_add.items():\n",
    "        \n",
    "            print (domain_name)\n",
    "            print (to_add)\n",
    "            seqs_to_add = []\n",
    "            \n",
    "\n",
    "            original_seqs = sequence.readFastaFile(f'{profile_name}/{domain_name}.fasta')\n",
    "\n",
    "            add_short = [\" \".join(x.split(\"information_\")[1].split(\"_\")[0:2]) for x in to_add]\n",
    "\n",
    "        \n",
    "            if len(add_short) > 1:\n",
    "                seq_txt = 'sequences'\n",
    "            else:\n",
    "                seq_txt = 'sequence'\n",
    "    \n",
    "\n",
    "            print (f'For domain {domain_name} we have to add {len(add_short)} {seq_txt}\\n')\n",
    "            print (f'Sequences where we found domains are - {\", \".join(add_short)}\\n')\n",
    "            for add in to_add:\n",
    "                   \n",
    "                \n",
    "                   start = region_dict[add][domain_name][0]\n",
    "                   end = region_dict[add][domain_name][1]\n",
    "                   \n",
    "                   # Add them to the profile dict\n",
    "                   profile_dict[domain_name].append(add)\n",
    "\n",
    "                   for seq in regions:\n",
    "                       if seq.name == add:\n",
    "                           seq_to_add = sequence.Sequence(\"\".join(seq.sequence[start:end]), name=f'{add}_{domain_name}')\n",
    "                           seqs_to_add.append(seq_to_add)\n",
    "\n",
    "\n",
    "            sequence.writeFastaFile(f'{profile_name}/{domain_name}.fasta', original_seqs + seqs_to_add)\n",
    "\n",
    "\n",
    "            # Delete the old alignment and HMM\n",
    "            os.remove(f'{profile_name}/{domain_name}.aln')\n",
    "            os.remove(f'{profile_name}/{domain_name}.hmm')\n",
    "\n",
    "            # Make new alignment\n",
    "            stdoutdata = subprocess.getoutput(f'mafft --genafpair --maxiterate 1000 --reorder {profile_name}/{domain_name}.fasta > {profile_name}/{domain_name}.aln')\n",
    "\n",
    "            # Make new HMM\n",
    "            stdoutdata = subprocess.getoutput(f'hmmbuild {profile_name}/{domain_name}.hmm {profile_name}/{domain_name}.aln')\n",
    "    return profile_dict\n",
    "\n",
    "                   \n",
    "def search_domains(base_dir, filename, file_dir, profile_name, count = 0):\n",
    "                   \n",
    "                   \n",
    "    # Load the tree        \n",
    "    tree = Tree(f'{filename}.nwk', format=1)\n",
    "\n",
    "    region_dict, domain_dict = search_for_domains(base_dir, file_dir, profile_name)\n",
    "    profile_dict = load_profile_dict(base_dir, profile_name)\n",
    "\n",
    "    domains_to_add = check_for_domains_to_add(profile_dict, domain_dict)\n",
    "                   \n",
    "    print ('This is the current tree \\n')\n",
    "                   \n",
    "    output_tree = display_tree(tree, region_dict, profile_name)\n",
    "\n",
    "\n",
    "    if domains_to_add:\n",
    "        profile_dict = add_domains(domains_to_add, region_dict, profile_dict, profile_name)\n",
    "        \n",
    "        count +=1\n",
    "        new_profile_dict_name = f'{profile_name}_{filename}_{count}'\n",
    "        save_new_profile_dict(profile_dict, profile_name, new_profile_dict_name)\n",
    "                   \n",
    "        if not os.path.isdir(f'{base_dir}/profiles/{new_profile_dict_name}'):\n",
    "\n",
    "                   \n",
    "            shutil.copytree(f'{profile_name}', f'{base_dir}/profiles/{new_profile_dict_name}')\n",
    "\n",
    "        search_domains(base_dir, filename, file_dir, f'{profile_name}_{filename}_{count}', count)\n",
    "\n",
    "    else:\n",
    "        print ('No new domains discovered')\n",
    "\n",
    "\n",
    "                   \n",
    "\n",
    "def save_new_profile_dict(profile_dict, profile_name, new_profile_dict_name):\n",
    "    remove_temp_files(profile_name + \"/*.p\")\n",
    "                   \n",
    "\n",
    "    with open(f'{profile_name}/{new_profile_dict_name}.p', 'wb') as profile_path:\n",
    "                   pickle.dump(profile_dict, profile_path)\n",
    "                   \n",
    "\n",
    "\n",
    "def remove_temp_files(path):\n",
    "    files = glob.glob(path)\n",
    "    for f in files:\n",
    "        os.remove(f)\n",
    "                   \n",
    "def display_tree(tree, region_dict, filename):\n",
    "\n",
    "    # Save the coloured tips to file\n",
    "    output_tree, ts = colour_tips(tree, tag_dict, colour_dict, region_dict, outpath= f'tree_output/{filename}.png', custom_layout=False)\n",
    "                                      \n",
    "    return (output_tree, ts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main method for running the workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Making directory\n",
      "\n",
      "Creating alignment\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sequence\n",
    "import shutil\n",
    "import subprocess\n",
    "import shutil\n",
    "import glob\n",
    "import pickle\n",
    "from collections import defaultdict\n",
    "from Bio import SearchIO\n",
    "from ete3 import Tree, TreeStyle, TextFace, add_face_to_node, SeqMotifFace, NodeStyle, faces, ImgFace, CircleFace, AttrFace\n",
    "\n",
    "# Create and setup the working directory\n",
    "file_dir = setup_working_dir(base_dir, filename)\n",
    "\n",
    "# Move into the working directory\n",
    "os.chdir(file_dir)\n",
    "\n",
    "# Load the sequences\n",
    "regions = sequence.readFastaFile(f'{filename}.fasta')\n",
    "\n",
    "# If we haven't already, make an alignment a tree and place the outgroup correctly\n",
    "\n",
    "make_alignment(filename)\n",
    "make_tree(filename)\n",
    "place_outgroup(filename, outgroup)\n",
    "\n",
    "# Search for domains in the sequences\n",
    "output_tree = search_domains(base_dir, filename, file_dir, profile_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
